src/useStorageState/index.ts

```diff
@@ -1,20 +1,44 @@
 import { useState } from 'react';
 
-function useStorageState<T>(storage: Storage, key: string, defaultValue?: T | (() => T)) {
-  const [state, setState] = useState<T | undefined>(() => {
-    const raw = storage.getItem(key);
-    if (raw !== null) {
-      return JSON.parse(raw);
-    }
-    if (typeof defaultValue === 'function') {
-      return (defaultValue as () => T)();
-    }
-    return defaultValue;
-  });
-  function updateState(value?: T) {
+interface IInitFunc<T> {
+  (): T;
+}
+
+interface IFuncUpdater<T> {
+  (previousState?: T): T;
+}
+
+function isFunction<T>(obj: any): obj is T {
+  return typeof obj === 'function';
+}
+
+function getByKey<T>(
+  storage: Storage,
+  key: string,
+  defaultValue?: T | IInitFunc<T>,
+): T | undefined {
+  const raw = storage.getItem(key);
+  if (raw) {
+    return JSON.parse(raw);
+  }
+  if (isFunction<IInitFunc<T>>(defaultValue)) {
+    return defaultValue();
+  }
+  return defaultValue;
+}
+
+function useStorageState<T>(storage: Storage, key: string, defaultValue?: T | IInitFunc<T>) {
+  const [state, setState] = useState<T | undefined>(() => getByKey(storage, key, defaultValue));
```

Comment:
感觉做成闭包函数更简单也更清晰一些。。没必要多些那么多手动推导的类型。。

```suggestion
  const [state, setState] = useState<T | undefined>(() => getByKey(storage, key, defaultValue));
  function getByKey() {
    const raw = storage.getItem(key);
    if (raw) {
      return JSON.parse(raw);
    }
    if (typeof defaultValue === 'function') {
      return (defaultValue as () => T)();
    }
    return defaultValue;
  }
```

Reply:
思路好，我回去再改改

Reply:
改好了